balancer.go

```diff
@@ -0,0 +1,329 @@
+			},
+			MinVersion:       tls.VersionTLS13,
+			MaxVersion:       tls.VersionTLS13,
+			CurvePreferences: []tls.CurveID{tls.CurveP521, tls.CurveP384, tls.CurveP256},
+		}
+
+		scheduleListeners[i] = schedule{port, config, identity}
+		i++
+	}
+
+	// Schedule listeners one by one and fail if any of them fail
+	derCtx, derCancel := context.WithCancel(lb.runCtx)
+	errChan := make(chan error, len(scheduleListeners))
+	defer close(errChan)
+
+	wg := sync.WaitGroup{}
+	for _, params := range scheduleListeners {
+		wg.Add(1)
+		go func(ctx context.Context, cancelAll context.CancelFunc, errChan chan error, toSchedule schedule) {
+
+			// Don't forget to close all contexts
+			defer derCancel()
+			// Try to listen
+			listen, err := tls.Listen("tcp", fmt.Sprintf("localhost:%d", toSchedule.port), toSchedule.tls)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to listen on port")
+				wg.Done()
+				return
+			}
+
+			// Spawn the coroutine to watch for the context break
+			go func(l net.Listener) {
+				// Here all required procedures were established for the listener thread
+				wg.Done()
+				// Await for context break
+				<-ctx.Done()
+				err := listen.Close()
+				lb.logger.Info().Msgf("closing listener at port %d", toSchedule.port)
+				if err != nil {
+					lb.logger.Err(err).Msgf("error closing listener at port %d", toSchedule.port)
+				}
+			}(listen)
+
+			lb.logger.Info().Msgf("listening at port %d", toSchedule.port)
+
+			// Bind pool and forwarder
+			currentPool := toSchedule.pool
+			var forwarder *Forwarder
+
+			// Maintain incoming connections
+			for {
+
+				conn, err := listen.Accept()
+				if err != nil {
+					if strings.Contains(err.Error(), "closed network") {
+						// Graceful
+						lb.logger.Debug().Msgf("listener closed, closed network for port: %d", toSchedule.port)
+						errChan <- nil
+					} else {
+						// Other kind of error
+						lb.logger.Err(err).Msgf("failed to accept connection, error")
+						errChan <- fmt.Errorf("failed to listen on port, error: %w", err)
+					}
+					lb.mutex.Lock()
+					delete(lb.forwarderMap, currentPool.Identity())
+					lb.mutex.Unlock()
+					return
+				}
+
+				lb.logger.Debug().Msgf("accepting request for port %d", toSchedule.port)
+
+				// Convert to TLS and proceed with the handshake
+				tlsConn := conn.(*tls.Conn)
+				err = tlsConn.Handshake()
+				if err != nil {
+					lb.logger.Err(err).Msg("cannot complete handshake")
+					err = tlsConn.Close()
+					if err != nil {
+						lb.logger.Err(err).Msg("cannot close connection after failed handshake")
+					}
+					continue
+				}
+
+				// Verify certificate CN and find or create corresponding backend to dispatch
+				certs := tlsConn.ConnectionState().PeerCertificates
+				if len(certs) == 0 {
+					lb.logger.Error().Msg("failed to extract certificate")
+					err = tlsConn.Close()
+					if err != nil {
+						lb.logger.Err(err).Msg("cannot close connection after certificate failure")
+					}
+					continue
+				}
```

Comment:
I would suggest to split this `Listen()` function into smaller function line:
 authConn(conn) 
handleClientConn 
forward(conn, ..)

Reply:
I agree with your suggestion. However, I'm not sure I'd have enough time for this split today, and as well there will be two additional PRs to deal with.